### PR TITLE
Add `mosaicml/llm-foundry` Docker workflow

### PR DESCRIPTION
**Pull request by abhi-mosaic**
@mvpatel2000 will need your guidance on how to test this :) 

Basically I'd like any new push to `main` to produce new Docker image tags under `mosaicml/llm-foundry:[torch_cuda]-[commithash_or_tag]`. The public DockerHub will be here: https://hub.docker.com/r/mosaicml/llm-foundry

The intention is for these images to have the **dependencies** for LLM-Foundry but not install the package itself. So it's really fast to `pip install` a custom git branch of LLM Foundry or `pip install` a version from PyPi, but there will be no package to begin with. This will prevent user errors where you might have conflicting llmfoundry packages on your system or venv.

### CODE REVIEW

**Comment by abhi-mosaic** [comment]
Confirmed that this works!
```
integrations:
- integration_type: git_repo
  git_repo: mosaicml/llm-foundry
  git_branch: abhi/docker # use your branch
  # git_commit: # OR use your commit hash
  pip_install: -e .[gpu]
  ssh_clone: false # Should be true if using a private repo

# We are fetching, converting, and training on the 'val' split
# as it is small and quick to get going for this demo.
# For real training runs, follow the instructions in `llm-foundry/scripts/train/README.md`
# to convert and host the full 'train' dataset.
command: |
  cd llm-foundry/scripts
  python data_prep/convert_dataset_hf.py \
    --dataset c4 --data_subset en \
    --out_root ./my-copy-c4 --splits train_small val_small \
    --concat_tokens 2048 --tokenizer EleutherAI/gpt-neox-20b --eos_text '<|endoftext|>'
  composer train/train.py train/yamls/pretrain/mpt-1b.yaml \
    train_loader.dataset.split=train_small \
    eval_loader.dataset.split=val_small \
    max_duration=100ba \
    eval_interval=0
# image: mosaicml/pytorch:1.13.1_cu117-python3.10-ubuntu20.04
image: mosaicml/llm-foundry:1.13.1_cu117-latest
# Mosaic Cloud will use run_name (with a unique suffix) to populate the env var $RUN_NAME
run_name: mosaic-gpt-1b-gpus-8

gpu_num: 8
gpu_type: a100_80gb
cluster: # replace with your cluster here!

# The below is injected as a YAML file: /mnt/config/parameters.yaml
# but is not used in this example.
parameters: {}
```

File: mcli/mcli-hf-eval.yaml [hunk 1]
```diff
@@ -46,7 +46,7 @@ parameters:
   model:
     name: hf_causal_lm
     pretrained_model_name_or_path: ${model_name_or_path}
-    device: cpu
+    init_device: cpu
```
Comment (mvpatel2000) [inline mcli/mcli-hf-eval.yaml:49]: along for ride?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# CCopyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+ARG BASE_IMAGE
+FROM $BASE_IMAGE
+
+
+RUN git clone -b main https://github.com/mosaicml/llm-foundry.git && \
+    pip install --no-cache-dir "./llm-foundry[gpu]" && \
```
Comment (mvpatel2000) [inline Dockerfile:9]:
```suggestion

# Install and uninstall foundry to cache foundry requirements
RUN git clone -b main https://github.com/mosaicml/llm-foundry.git && \
    pip install --no-cache-dir "./llm-foundry[gpu]" && \
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# CCopyright 2022 MosaicML LLM Foundry authors
```
Comment (mvpatel2000) [inline Dockerfile:1]:
```suggestion
# Copyright 2022 MosaicML LLM Foundry authors
```

**Review by mvpatel2000 — APPROVED** [review]
LGTM!

**Pull request merged**